File: JobTracker/Models/Lead.swift
//
//  Lead.swift
//  JobTracker
//
//  A manually-added prospect: a job posting to apply to, or a person to
//  reach out to. Lives in its own "Leads" section until acted on.
//

import Foundation
import SwiftData

enum LeadType: String, Codable, CaseIterable, Identifiable {
    case jobPosting
    case person

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .jobPosting: return "Job Posting"
        case .person:     return "Person / Contact"
        }
    }

    var systemImage: String {
        switch self {
        case .jobPosting: return "link"
        case .person:     return "person.crop.circle"
        }
    }
}

enum LeadStage: String, Codable, CaseIterable, Identifiable {
    case todo
    case contacted
    case applied
    case done

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .todo:      return "To Do"
        case .contacted: return "Contacted"
        case .applied:   return "Applied"
        case .done:      return "Done"
        }
    }

    var color: String {
        switch self {
        case .todo:      return "orange"
        case .contacted: return "teal"
        case .applied:   return "blue"
        case .done:      return "green"
        }
    }
}

@Model
final class Lead {
    var id: UUID = UUID()
    var title: String = ""
    var company: String = ""
    var url: String = ""
    var personName: String = ""
    var typeRaw: String = LeadType.jobPosting.rawValue
    var stageRaw: String = LeadStage.todo.rawValue
    var notes: String = ""
    var createdDate: Date = Date()
    var lastUpdated: Date = Date()

    init(type: LeadType = .jobPosting) {
        self.id = UUID()
        self.typeRaw = type.rawValue
        self.createdDate = Date()
        self.lastUpdated = Date()
    }

    var type: LeadType {
        get { LeadType(rawValue: typeRaw) ?? .jobPosting }
        set { typeRaw = newValue.rawValue }
    }

    var stage: LeadStage {
        get { LeadStage(rawValue: stageRaw) ?? .todo }
        set { stageRaw = newValue.rawValue }
    }

    var linkURL: URL? {
        guard !url.isEmpty else { return nil }
        return URL(string: url.hasPrefix("http") ? url : "https://\(url)")
    }
}

File: JobTracker/Models/Lead.cs
// A manually-added prospect: a job posting to apply to, or a person to
// reach out to. Lives in its own "Leads" section until acted on.

namespace JobTracker.Models;

public enum LeadType { JobPosting, Person }

public enum LeadStage { Todo, Contacted, Applied, Done }

public static class LeadExtensions
{
    public static string DisplayName(this LeadType type) => type switch
    {
        LeadType.JobPosting => "Job Posting",
        LeadType.Person => "Person / Contact",
        _ => type.ToString(),
    };

    public static string DisplayName(this LeadStage stage) => stage switch
    {
        LeadStage.Todo => "To Do",
        LeadStage.Contacted => "Contacted",
        LeadStage.Applied => "Applied",
        LeadStage.Done => "Done",
        _ => stage.ToString(),
    };

    public static string ColorKey(this LeadStage stage) => stage switch
    {
        LeadStage.Todo => "Orange",
        LeadStage.Contacted => "Teal",
        LeadStage.Applied => "Blue",
        LeadStage.Done => "Green",
        _ => "Gray",
    };
}

public class Lead
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Title { get; set; } = "";
    public string Company { get; set; } = "";
    public string Url { get; set; } = "";
    public string PersonName { get; set; } = "";
    public string TypeRaw { get; set; } = LeadType.JobPosting.ToString();
    public string StageRaw { get; set; } = LeadStage.Todo.ToString();
    public string Notes { get; set; } = "";
    public DateTimeOffset CreatedDate { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset LastUpdated { get; set; } = DateTimeOffset.UtcNow;

    public Lead() { }

    public Lead(LeadType type = LeadType.JobPosting)
    {
        TypeRaw = type.ToString();
    }

    public LeadType Type
    {
        get => Enum.TryParse<LeadType>(TypeRaw, out var v) ? v : LeadType.JobPosting;
        set => TypeRaw = value.ToString();
    }

    public LeadStage Stage
    {
        get => Enum.TryParse<LeadStage>(StageRaw, out var v) ? v : LeadStage.Todo;
        set => StageRaw = value.ToString();
    }

    public Uri? LinkUrl
    {
        get
        {
            if (string.IsNullOrEmpty(Url)) return null;
            var value = Url.StartsWith("http", StringComparison.OrdinalIgnoreCase) ? Url : $"https://{Url}";
            return Uri.TryCreate(value, UriKind.Absolute, out var uri) ? uri : null;
        }
    }
}
